library/alloc/src/collections/btree/node.rs

```diff
@@ -280,26 +279,28 @@ unsafe impl<K: Send, V: Send, Type> Send for NodeRef<marker::Owned, K, V, Type>
 impl<BorrowType, K, V> NodeRef<BorrowType, K, V, marker::Internal> {
     /// Exposes the data of an internal node for reading.
     ///
-    /// Returns a raw ptr to avoid invalidating other references to this node,
-    /// which is possible when BorrowType is marker::ValMut.
+    /// Returns a raw ptr to avoid invalidating other references to this node.
     fn as_internal_ptr(&self) -> *const InternalNode<K, V> {
         self.node.as_ptr() as *const InternalNode<K, V>
     }
 }
 
 impl<'a, K, V> NodeRef<marker::Mut<'a>, K, V, marker::Internal> {
-    /// Exposes the data of an internal node for reading,
-    /// when we know we have exclusive access.
-    fn as_internal(&mut self) -> &InternalNode<K, V> {
+    /// Exposes the data of an internal node for reading.
+    ///
+    /// # Safety
+    /// We have exclusive access to the node (and not just to the `NodeRef`).
```

Comment:
Isn't there supposed to be an invariant that a `NodeRef<marker::Mut` always has exclusive access to the node?

Reply:
Somewhat, but with at least these exceptions:
- temporary ones with `reborrow_mut`
- temporary one with `ptr::read` in `remove_kv_tracking` (I have a PR queued to use `reborrow_mut` instead)
- just realized today, the pointer returned by `insert_fit`

Reply:
> temporary ones with reborrow_mut

Oh, reborrow_mut doesn't use some lifetimes to make sure only one of the two ptrs can be used?
But then I'd say it is reborrow_mut that should be unsafe, with the safety invariant being "do not use the original ref until you stopped using the reborrowed one". (Somewhat like dorman-mut-ref actually...)

Reply:
Well  `reborrow_mut` _is_ unsafe, with ample warning, and I think the borrow checker makes sure you can only use the top one, but you can still use it to deallocate the node that the original one refers to, and that's not cool…